charms/knative-operator/src/charm.py

```diff
@@ -33,28 +35,70 @@ def __init__(self, *args):
 
         self._app_name = self.model.app.name
         self._namespace = self.model.name
-        self.resource_handler = KRH(
-            template_files=self._template_files,
-            context=self._context,
-            field_manager=self._namespace,
+        self._src_dir = Path("src")
+        self._template_files_ext = None
+        self._resource_handler = None
+
+        self._scraping = MetricsEndpointProvider(
+            self,
+            relation_name="metrics-endpoint",
+            jobs=[{"static_configs": [{"targets": [f"{self._otel_exporter_ip}:8889"]}]}],
         )
         self._operator_service = "/ko-app/operator"
         self._container = self.unit.get_container(self._app_name)
         for event in [self.on.install, self.on.config_changed]:
             self.framework.observe(event, self._main)
+
+        self.framework.observe(
+            self.on["otel-collector"].relation_changed, self._on_otel_collector_relation_changed
+        )
         self.framework.observe(
             self.on.knative_operator_pebble_ready,
             self._on_knative_operator_pebble_ready,
         )
         self.framework.observe(self.on.remove, self._on_remove)
 
+    @property
+    def resource_handler(self):
+        """Returns an instance of the KubernetesResourceHandler with initial parameters."""
+        if not self._resource_handler:
+            self._resource_handler = KRH(
+                template_files=self._template_files,
+                context=self._context,
+                field_manager=self._namespace,
+            )
+        return self._resource_handler
+
     @property
     def _template_files(self):
-        src_dir = Path("src")
-        manifests_dir = src_dir / "manifests"
+        manifests_dir = self._src_dir / "manifests"
         eventing_manifests = [file for file in glob.glob(f"{manifests_dir}/*.yaml.j2")]
+        # Extend the list of template files if needed
+        # self._template_files_ext should be a list
+        if self._template_files_ext:
+            eventing_manifests.extend(self._template_files_ext)
         return eventing_manifests
 
+    @property
+    def _otel_exporter_ip(self):
+        """Returns the ClusterIP of the otel-export service."""
+        try:
+            exporter_service = self.resource_handler.lightkube_client.get(
+                res=Service, name="otel-export", namespace=self._namespace
+            )
+            exporter_ip = exporter_service.spec.clusterIP
+        except ApiError as e:
+            if e.status.code == 404:
+                logger.info(
+                    "The OpenTelemetry Collector may not be deployed yet."
+                    "This may be temporary or due to a missing otel-collector relation."
+                )
+                return ""
+            logger.error(traceback.format_exc())
+            raise
+        else:
+            return exporter_ip
```

Comment:
Can we add a unit test for this that tries `_otel_exporter_ip()` when:
* lightkube_client.get() returns a service
* lightkube_client.get() raises a 404 error
* lightkube_client.get() raises some other error
?

Reply:
I believe this is still unresolved

Reply:
I have added a test case for it. Let me know if that suffices.

Reply:
perfect!